lib/MemoryModel/LocationSet.cpp

```diff
@@ -55,12 +55,74 @@ bool LocationSet::isConstantOffset() const
-/// v1*sz(t2) + v2*sz(t3) + v3 
-s64_t LocationSet::accumulateConstantOffset() const{
+/// Return element number of a type
+/// (1) StructType, return flatterned number elements, i.e., the index of the last element (getFlattenedOffsetVec().back()) plus one.
+/// (2) ArrayType, return number of elements
+/// (3) PointerType, return the element number of the pointee 
+/// (4) non-pointer SingleValueType, return 1
+u32_t LocationSet::getElementNum(const Type* type) const{
+    u32_t sz = 1;
+    if(const ArrayType* aty = SVFUtil::dyn_cast<ArrayType>(type))
+    {
+        /// handle nested arrays
+        const Type* innerTy = aty;
+        while (const ArrayType* arr = SVFUtil::dyn_cast<ArrayType>(innerTy))
+        {
+            sz *= arr->getNumElements();
+            innerTy = arr->getElementType();
+        }
+    }
+    else if (const StructType *sty = SVFUtil::dyn_cast<StructType>(type) )
+    {
+        const vector<u32_t> &so = SymbolTableInfo::SymbolInfo()->getFlattenedOffsetVec(sty);
+        sz = so.back() + 1;
+    }
+    else if (type->isSingleValueType())
+    {
+        /// This is a pointer arithmic
+        if(const PointerType* pty = SVFUtil::dyn_cast<PointerType>(type) ){
+            sz = getElementNum(pty->getElementType());
+        }
+    }
+    else{
+        SVFUtil::outs() << "GepIter Type" << *type << "\n";
+        assert(false && "What other types for this gep?");
+    }
+    return sz;
+}
 
-    return 0;
+/// Return accumulated constant offset
+/// 
+/// "value" is the offset variable (must be a constant)
+/// "type" is the location where we want to compute offset
+/// e.g., %3 = getelementptr inbounds [10 x i32], [10 x i32]* %1, i64 0, i64 5 
+/// offsetValues[0] value: i64 0, type: [10 x i32]*
+/// offsetValues[1] value: i64 5, type: [10 x i32]
+/// 
+/// Given a vector: [(v1,t1), (v2,t2), (v3,t3)]
+/// totalConstOffset = v1 * sz(t2) + v2 * sz(t3) + v3 * 1
+/// If the vector only has one element (one gep operand), then it must be a pointer arithmetic and type must be a PointerType
+/// totalConstOffset = v1 * sz(t1) 
+s64_t LocationSet::accumulateConstantOffset() const{
+    
+    assert(isConstantOffset() && "not a constant offset");
+
+    s64_t totalConstOffset = 0;
+    u32_t sz = 1;
+    for(int i = offsetValues.size() - 1; i >= 0; i--){
+        const Value* value = offsetValues[i].first;
+        const Type* type = offsetValues[i].second;
+        const ConstantInt *op = SVFUtil::dyn_cast<ConstantInt>(value);
+        assert(op && "not a constant offset?");
+        /// if this gep only has one operand, the gepIterType must be the pointer type, and we will need to retrieve size of its elementType.
+        if(offsetValues.size()==1){
+            assert(SVFUtil::isa<PointerType>(type) && "If gep has only one operand, its gepIterType must be PointerType!");
+            sz = getElementNum(type);
+        }
+        totalConstOffset += op->getSExtValue() * sz; 
+        sz *= getElementNum(type);
```

Comment:
Is this accumulation necessary, since getElementNum handles flattening?

Reply:
The flattening mechanism builds a map via `const vector<u32_t> &flattenVec = SymbolTableInfo::SymbolInfo()->getFlattenedOffsetVec(structTy)` to map a field index (from a gep instruction) to a flattened index of a struct type `structTy`. 

`j = flattenVec[I];` // `i` is the field index/offset in the gep instruction  and `j` is the index after flattening `structTy`.

For example
````
struct inner_struct{
int x;
int y;
}
struct structTy{
int a[10];
inner_struct b;
float c;
};

%q = getelementptr %structTy, %structTy* %p, i32 0, i32 2
````

The field index is 2 for accessing `c` in `structTy` at this gep instruction. `flattenVec[2]` returns `12=10+2`.

However, `getFlattenedOffsetVec` is only used to flatten structs for now. We will need `getElementNum` to calculate the sizes for all other types, and also recursively calculate the size of pointee if the type is a PointerType.

Reply:
Oh sorry, I meant the accumulation with respect to `sz` (`*=`). For example, in an array of array of array, we have the size of the first (encompassing itself, the second, and the third), times the second (encompassing itself and the third), time the third (encompassing itself). (May be off by one in the example.)

Reply:
I didn't quite get that. `sz *= getElementNum(type)` is to calculate the size for the next loop iteration, based on the equation at line 103.

Reply:
`getElementNum(t)` accounts for flattening, so I may be misunderstanding why `sz` is multiplied rather than just assigned.